engine/dlib/src/dlib/profile/profile.cpp

```diff
@@ -14,52 +14,563 @@
+// Properties
+
+static void CreateProperty(ProfileProperty* prop)
+{
+    for (ProfileListener* p = g_ProfileListeners; p!=0; p = p->m_Next)
+    {
+        CreateProperty(p, prop);
+    }
+}
 
-// end unit test
+static void CreateRegisteredProperties()
+{
+    DM_MUTEX_SCOPED_LOCK(g_ProfileLock);
+
+    for (ProfileListener* p = g_ProfileListeners; p!=0; p = p->m_Next)
+    {
+        CreateListenerProperties(p);
+    }
+}
+
+// Invoked when first property is initialized
+static void PropertyInitialize()
+{
+    if (dmAtomicIncrement32(&g_PropertyInitialized) != 0)
+        return;
+
+    g_ProfileLock = dmMutex::New();
+
+    g_ProfileProperties[0].m_Name = "Root";
+    g_ProfileProperties[0].m_NameHash = dmHashString32(g_ProfileProperties[0].m_Name);
+    g_ProfileProperties[0].m_Type = PROFILE_PROPERTY_TYPE_GROUP;
+    g_ProfileProperties[0].m_ParentIdx = PROFILE_PROPERTY_INVALID_IDX;
+    g_ProfilePropertyCount = 1;
+}
+
+static inline bool IsValidIndex(ProfileIdx idx)
+{
+    return idx != PROFILE_PROPERTY_INVALID_IDX;
+}
+
+static ProfileProperty* GetPropertyFromIdx(ProfileIdx idx)
+{
+    if (!IsValidIndex(idx))
+        return 0;
+    return &g_ProfileProperties[idx];
+}
+
+static ProfileProperty* AllocateProperty(ProfileIdx* idx)
+{
+    DM_MUTEX_SCOPED_LOCK(g_ProfileLock);
+    if (g_ProfilePropertyCount >= PROFILER_MAX_NUM_PROPERTIES)
+        return 0;
+    *idx = g_ProfilePropertyCount++;
+    return &g_ProfileProperties[*idx];
+}
+
+static void RegisterProperty(ProfileProperty* prop, ProfileIdx idx, const char* name, const char* desc, uint32_t flags, ProfileIdx parentidx)
+{
+    memset(prop, 0, sizeof(ProfileProperty));
+    prop->m_Name        = name;
+    prop->m_Desc        = desc;
+    prop->m_Flags       = flags;
+    prop->m_Idx         = idx; // perhaps a little redundant but nice when debugging the property
+    prop->m_ParentIdx   = parentidx;
+    prop->m_NameHash    = dmHashString64(name);
+    prop->m_FirstChild  = PROFILE_PROPERTY_INVALID_IDX;
+    prop->m_Sibling     = PROFILE_PROPERTY_INVALID_IDX;
+
+    // If this fails, then we've messed up the dependencies.
+    // We must register the parent first. See ALLOC_PROP_AND_CHECK()
+    assert(idx > parentidx);
+
+    ProfileProperty* parent = GetPropertyFromIdx(prop->m_ParentIdx);
+    if (parent)
+    {
+        prop->m_Sibling = parent->m_FirstChild;
+        parent->m_FirstChild = idx;
+    }
+}
+
+// This macro also resolves the parentfn() -> parentidx,
+// which allows us to create each parent before its children.
+#define ALLOC_PROP_AND_CHECK(_PARENTFN)                     \
+    PropertyInitialize();                                   \
+    ProfileIdx parentidx = (_PARENTFN) ? *(_PARENTFN)() : 0;\
+    ProfileIdx idx; \
+    ProfileProperty* prop = AllocateProperty(&idx); \
+    if (!prop) \
+        return PROFILE_PROPERTY_INVALID_IDX;
+
+
+ProfileIdx ProfileRegisterPropertyGroup(const char* name, const char* desc, ProfileIdx* (*parentfn)())
```

Comment:
A previous contributor fix, was to use an actual function to get (and initialize) the parent id.
This solves a dependency issue where they could be initialized in wrong order.